[WebXR] Display ARKit captured image in overlay view

djg opened the pull request:
#### f6e1ce804c6ed01924775d324fc2c745c27c4e14
<pre>
[WebXR] Display ARKit captured image in overlay view
<a href="https://bugs.webkit.org/show_bug.cgi?id=262773">https://bugs.webkit.org/show_bug.cgi?id=262773</a>
rdar://116502294

Reviewed by Dean Jackson.

Unlike other WebXR compositor systems, ARKit doesn&apos;t provide a framework for
rendering immersive AR content, just a series of captured camera images with
camera transformation, so we need to implement our own compositor.

This patch introduces WKARPresentationSession that is responsible for managing a
view and compositing the captured camera images with WebGL rendering. On
entering an immersive AR session, WKARPresentationSession creates a new
UIViewController and makes it the presenting view controller for UIKit. The
camera images, which are provided in CVPixelBuffers, are displayed by setting
them as the CALayer content of a UIView. In a following patch, the WebGL content
will be composited on top of the camera image.

Update of the captured camera image is driven off the main thread using the
setting of the animation frame callback from the page to trigger the update.

It is intended that the view should be fullscreen but currently it is rendered
as a small view to allow exiting an immersive AR session by touching outside the
view to dismiss it.

* Source/WebKit/Shared/Cocoa/ARKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/ARKitSoftLink.mm:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.h:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
(WebKit::ARKitCoordinator::startSession):
(WebKit::ARKitCoordinator::endSessionIfExists):
(WebKit::ARKitCoordinator::scheduleAnimationFrame):
(WebKit::ARKitCoordinator::createSessionIfNeeded):
(WebKit::ARKitCoordinator::renderLoop):
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.h: Copied from Source/WebKit/Shared/Cocoa/ARKitSoftLink.h.
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm: Added.
(-[WKARPresentationSessionDescriptor copyWithZone:]):
(-[WKARPresentationSessionDescriptor presentingViewController]):
(-[WKARPresentationSessionDescriptor setPresentingViewController:]):
(-[_WKARPresentationSession initWithSession:descriptor:]):
(-[_WKARPresentationSession session]):
(-[_WKARPresentationSession startFrame]):
(-[_WKARPresentationSession present]):
(-[_WKARPresentationSession terminate]):
(-[_WKARPresentationSession loadView]):
(-[_WKARPresentationSession viewDidLoad]):
(-[_WKARPresentationSession preferStatusBarHidden]):
(-[_WKARPresentationSession viewWillAppear:]):
(-[_WKARPresentationSession viewWillDisappear:]):
(-[_WKARPresentationSession _enterFullscreen]):
(createPresesentationSession):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Originally-landed-as: 269064@main (6371344eac3b). rdar://116502294
Canonical link: <a href="https://commits.webkit.org/269352@main">https://commits.webkit.org/269352@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9cfe6b75126afb43817141a101c400f7d0285b10

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/22318 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/22517 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/23393 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/24218 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/20656 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/26779 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/22847 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/24218 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/22554 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/26779 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/23393 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/25073 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/26779 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/23393 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/25073 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/26779 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/23393 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/25073 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/20987 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/22847 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/20445 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/23393 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5313 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/24675 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/20983 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -0,0 +1,204 @@
+
+@end
+
+id<WKARPresentationSession> createPresesentationSession(ARSession *session, WKARPresentationSessionDescriptor *descriptor)
+{
+    return [[_WKARPresentationSession alloc] initWithSession:session descriptor:descriptor];
+}
```
Switching to soft-linking of ARKit means that I couldn't add an extension category to ARSession. I've replaced the method I added to ARSession with a freestanding function.

grorg commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -0,0 +1,204 @@
+#if USE(ARPRESENTATIONSESSION_NOISY_LOG)
+#define NOISY_RELEASE_LOG(channel, ...) RELEASE_LOG(channel, __VA_ARGS__)
+#else
+#define NOISY_RELEASE_LOG(channel, ...) ((void)0)
```
Is the idea that these logs will be on in release once everything is "done"?

I think I'm wondering why you don't have noisy logging for debug builds using the normal `LOG` mechanism? Or do you need to build Release?

djg replied:
My intention is that I'll turn them off once everything is done, and they're just for me while I get through debugging without an Xcode that was attaching properly. (I've been using release builds for perf)

grorg commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -0,0 +1,204 @@
+    RetainPtr<CVPixelBufferRef> _capturedImage;
+}
+
+- (nonnull instancetype)initWithSession:(nonnull ARSession*)session descriptor:(nonnull WKARPresentationSessionDescriptor*)descriptor
```
You don't need the `nonnull` here (and you do possibly return null).

grorg commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -0,0 +1,204 @@
+    WeakObjCPtr<UIViewController> _presentingViewController;
+}
+
+- (instancetype)copyWithZone:(nullable NSZone *)zone
```
I don't think you need the `nullable` here.

djg replied:
Compiler disagrees: `Nullability specifier 'nonnull' conflicts with existing specifier 'nullable'`

grorg commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -0,0 +1,204 @@
+    return descriptor;
+}
+
+- (nullable UIViewController*)presentingViewController
```
No `nullable` here either.

grorg replied:
And other places.

djg replied:
Compiler disagrees: `Nullability specifier 'nonnull' conflicts with existing specifier '_Nullable'`

grorg commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -0,0 +1,204 @@
+    return self;
+}
+
+#pragma mark - WKARPresentationSession
```
Nit: Blank line after this.

grorg commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -0,0 +1,204 @@
+    RELEASE_LOG(XR, "%s", __FUNCTION__);
+}
+
+#pragma mark - UIViewController
```
Nit: Blank line after this.

grorg commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -0,0 +1,204 @@
+- (void)loadView
+{
+    RELEASE_LOG(XR, "%s", __FUNCTION__);
+    UIView *view = [[UIView alloc] initWithFrame:UIScreen.mainScreen.bounds];
```
Don't we want to retain this?

djg committed:
[WebXR] Display ARKit captured image in overlay view
https://bugs.webkit.org/show_bug.cgi?id=262773
rdar://116502294

Reviewed by Dean Jackson.

Unlike other WebXR compositor systems, ARKit doesn't provide a framework for
rendering immersive AR content, just a series of captured camera images with
camera transformation, so we need to implement our own compositor.

This patch introduces WKARPresentationSession that is responsible for managing a
view and compositing the captured camera images with WebGL rendering. On
entering an immersive AR session, WKARPresentationSession creates a new
UIViewController and makes it the presenting view controller for UIKit. The
camera images, which are provided in CVPixelBuffers, are displayed by setting
them as the CALayer content of a UIView. In a following patch, the WebGL content
will be composited on top of the camera image.

Update of the captured camera image is driven off the main thread using the
setting of the animation frame callback from the page to trigger the update.

It is intended that the view should be fullscreen but currently it is rendered
as a small view to allow exiting an immersive AR session by touching outside the
view to dismiss it.

* Source/WebKit/Shared/Cocoa/ARKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/ARKitSoftLink.mm:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.h:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
(WebKit::ARKitCoordinator::startSession):
(WebKit::ARKitCoordinator::endSessionIfExists):
(WebKit::ARKitCoordinator::scheduleAnimationFrame):
(WebKit::ARKitCoordinator::createSessionIfNeeded):
(WebKit::ARKitCoordinator::renderLoop):
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.h: Copied from Source/WebKit/Shared/Cocoa/ARKitSoftLink.h.
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm: Added.
(-[WKARPresentationSessionDescriptor copyWithZone:]):
(-[WKARPresentationSessionDescriptor presentingViewController]):
(-[WKARPresentationSessionDescriptor setPresentingViewController:]):
(-[_WKARPresentationSession initWithSession:descriptor:]):
(-[_WKARPresentationSession session]):
(-[_WKARPresentationSession startFrame]):
(-[_WKARPresentationSession present]):
(-[_WKARPresentationSession terminate]):
(-[_WKARPresentationSession loadView]):
(-[_WKARPresentationSession viewDidLoad]):
(-[_WKARPresentationSession preferStatusBarHidden]):
(-[_WKARPresentationSession viewWillAppear:]):
(-[_WKARPresentationSession viewWillDisappear:]):
(-[_WKARPresentationSession _enterFullscreen]):
(createPresesentationSession):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Originally-landed-as: 269064@main (6371344eac3b). rdar://116502294
Canonical link: https://commits.webkit.org/269352@main
